eng/scripts/Verify-And-Merge-PRs.ps1

```diff
@@ -1,111 +1,250 @@
+    return
   }
+
+  MergePRs $mergeable
 }
-while (($unMergablePRs.Count -gt 0) -and ($RetryCount -gt 0) -and $ShouldMerge)
 
+function ProcessPRMergeStatuses([array]$prData) {
+  for ($retry = 1; $retry -le 5; $retry++)
+  {
+    $curr, $prData = $prData, @()
+    foreach ($pr in ($curr | Where-Object { $_.Retry -ne $false }))
+    {
+      $prData += GetOrSetMergeablePR -repoOwner $pr.RepoOwner -repoName $pr.RepoName -prNumber $pr.Number
+    }
+
+    if ($SkipMerge -or $prData.Retry -notcontains $true) {
+      break
+    }
+
+    $sleep = [Math]::Pow(2, $retry)
+    LogInfo "Some PRs were not in a mergeable state, retrying after $sleep seconds..."
+    Start-Sleep -Seconds $sleep
+  }
 
-if (!$ReadyForMerge) {
-  LogError "At least one sync PR is not able to be merged please investigate and then retry running this job to auto-merge them again"
-  exit 1
+  if ($prData.Block -contains $true) {
+    LogError "The following sync PRs are not able to be merged. Investigate and then retry running this job to auto-merge them again"
+    $prData | Where-Object { $_.Block } | ForEach-Object { LogInfo $_.Url }
+    exit 1
+  }
+
+  return $prData
 }
 
-if ($ReadyForMerge -and $ShouldMerge)
-{
-  # Merge Pull Requests
-  foreach ($mergablePRObj in $mergablePRs)
+function MergePRs([array]$toMerge) {
+  foreach ($pr in $toMerge)
   {
-    $mergablePR = $mergablePRObj.Url
-    $mergeApiUrl = $mergablePR + "/merge"
-
-    Write-Host $mergablePRObj.HeadSHA
-    $data = @{
-      sha = $mergablePRObj.HeadSHA
+    LogInfo $toMerge.HeadSHA
+    $body = @{
+      sha = $toMerge.HeadSHA
       merge_method = "squash"
-    }
+    } | ConvertTo-Json -Compress
 
-    # Merge Pull Request
     try {
-      Write-Host "Merging $mergablePR"
-      $response = Invoke-RestMethod -Method Put -Headers $headers $mergeApiUrl -Body ($data | ConvertTo-Json)
+      LogInfo "Merging $($toMerge.MergeUrl)"
+      $response = Invoke-RestMethod -Method Put -Headers $Headers $toMerge.MergeUrl -Body $body
     }
     catch {
-      LogError "Invoke-RestMethod [$mergeApiUrl] failed with exception:`n$_"
-      exit 1
+      LogError "Invoke-RestMethod [$($toMerge.MergeUrl)] failed with exception."
+      throw
+    }
+  }
+}
+
+function RequestGithubGraphQL([string]$query, [object]$variables = @{}) {
+  $body = @{ query = $query; variables = $variables } | ConvertTo-Json -Depth 100
+
+  try {
+    $response = Invoke-RestMethod `
+      -Method Post `
+      -Uri "https://api.github.com/graphql" `
+      -Headers $Headers `
+      -Body $body `
+      -ContentType "application/json"
+  }
+  catch {
+    throw "Invoke-RestMethod failed for graphql operation: `n$query"
+  }
+
+  # The API will return 200 with a list of error messages for bad queries
+  if ($response.errors) {
+    throw $response.errors.message -join '`n'
+  }
+
+  return $response
+}
+
+function TryResolveAIReviewThreads([string]$repoOwner, [string]$repoName, [string]$prNumber) {
```

Comment:
Should we do this here or in our approve helper script?

Reply:
Good question. Seems like it does belong in the user approval stage instead. On the other hand, not everyone uses that script, so it could add more friction. If we want to make everyone use it, then I can also update the sync workflow doc to say you have to run the script.

Reply:
Yeah I think we probably should start having folks use it and I kind of feel like this belongs there, so things are resolved as the person not the bot.

Reply:
Ok I'll move it over.